MmSupervisorPkg/Core/Request/UpdateCommBuffer.c

```diff
@@ -0,0 +1,174 @@
+    goto Done;
+  }
+
+  // Okay, again, enough complaints, now get to work.
+  Status = ProcessUnblockPages (NewMemParam);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Failed to unblock the requested new communicate buffers - %r!\n", __FUNCTION__, Status));
+    goto Done;
+  }
+
+  Status = ProcessBlockPages (OldMemParam);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Failed to block memory %r!\n", __FUNCTION__, Status));
+    goto Done;
+  }
+
+Done:
+  return Status;
+}
+
+/**
+  Routine used to update communication buffers and core private mailbox region
+  to the update location. Given this routine could receive untrusted data, the
+  requested memory region has to be already mapped as "not present" prior to this
+  request and the received new buffer has to persist exactly the same memory
+  length and attribute as before. For requests that pass security checks, the new
+  region will be marked as R/W supervisor data page. The old buffers will be blocked
+  after this routine, and new return status will be populated to the new address.
+  The caller should be prepared to check data from newly supplied region upon returning.
+
+  @param[in]  UpdateCommBuffer  Input unblock parameters conveyed from non-MM environment
+
+  @retval EFI_SUCCESS             The requested region properly unblocked.
+  @retval EFI_ACCESS_DENIED       The request was made post lock down event.
+  @retval EFI_INVALID_PARAMETER   UnblockMemParams or its ID GUID is null pointer.
+  @retval EFI_SECURITY_VIOLATION  The requested region has illegal page attributes.
+  @retval EFI_OUT_OF_RESOURCES    The unblocked database failed to log new entry after
+                                  processing this request.
+  @retval Others                  Page attribute setting/clearing routine has failed.
+
+**/
+EFI_STATUS
+ProcessUpdateCommBufferRequest (
+  IN MM_SUPERVISOR_COMM_UPDATE_BUFFER  *UpdateCommBuffer
+  )
+{
+  EFI_STATUS                           Status = EFI_SUCCESS;
+  UINTN                                Index;
+  MM_SUPERVISOR_UNBLOCK_MEMORY_PARAMS  MmCoreDataDesc;
+
+  if ((mMmReadyToLockDone) || (AlreadyMoved)) {
+    // Note that this flag will be set once the policy is requested
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a - Comm buffer update requested after ready to lock, will not proceed!\n",
+      __FUNCTION__
+      ));
+    return EFI_ACCESS_DENIED;
+  }
+
+  // Some more sanity checks here
+  if (UpdateCommBuffer == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a - Invalid parameter detected - %p!\n", __FUNCTION__, UpdateCommBuffer));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Prepare the playground...
+  ZeroMem (&MmCoreDataDesc, sizeof (MM_SUPERVISOR_UNBLOCK_MEMORY_PARAMS));
+
+  // Now attempt to move the old pages to the newly supplied buffers
+  for (Index = 0; Index < MM_OPEN_BUFFER_CNT; Index++) {
+    CopyMem (&(MmCoreDataDesc.MemoryDescriptor), &(mMmSupervisorAccessBuffer[Index]), sizeof (EFI_MEMORY_DESCRIPTOR));

```

Comment:
I think the following would help reduce likelihood of the size getting out of sync if the type changes in the future.

```suggestion
    CopyMem (&(MmCoreDataDesc.MemoryDescriptor), &(mMmSupervisorAccessBuffer[Index]), sizeof (MmCoreDataDesc.MemoryDescriptor));
```